runner/local/queueing_runner.go

```diff
@@ -0,0 +1,339 @@
+package local
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/scootdev/scoot/runner"
+)
+
+// TODO(dbentley): should this be in queue.go
+// TODO(dbentley): should this whole package move to runner/runners?
+
+const QueueFullMsg = "No resources available. Please try later."
+const UnknownRunIdMsg = "Unknown run id."
+const RequestIsRunning = "Run %s is still running, please Abort it first."
+
+// QueueingRunner manages a queue of commands.
+// One request is run at a time, via an underlying delegate runner
+// (We expect, for now, the delegate runner will ultimately be a simpleRunner
+// QueueingRunner will reject requests if the queue is full (i.e., its length == its capacity)
+type QueueingRunner struct {
+	// channel for methods to send communication to eventLoop()
+	reqCh chan interface{}
+
+	// delegate that will run (we expect it to ultimately be a simpleRunner)
+	delegate    runner.Runner
+	capacity    int
+	nextRunId   int // used to assign unique runIds to the run requests
+	runnerAvail bool
+
+	// Runs are in one of three meta-states in QueueingRunner:
+	// *) queued
+	// *) errored (we couldn't delegate them)
+	// *) delegated
+
+	// q holds queued commands (and assigned ids)
+	q []commandAndId
+
+	// errored holds info for runs that have errored.
+	// We weren't able to send them to the delegate (either got an error from Run()
+	// or they were aborted before running
+	errored map[runner.RunId]runner.ProcessStatus
+
+	// delegated runs are stored in both qToDel and delToQ. Together, these
+	// form a BiMap (but Go doesn't make that easy), so we have two maps.
+	// qToDel maps IDs in our namespace to IDs in the delegate's namespace
+	// delToQ is vice versa
+	qToDel map[runner.RunId]runner.RunId
+	delToQ map[runner.RunId]runner.RunId
+}
+
+// NewQueueingRunner creates a new QueueingRunner.
+// delegate will be used to run.
+// capacity is the maximum entries to hold in the queue
+// runnerAvailableCh should send a struct whenever the runner is available
+func NewQueuingRunner(delegate runner.Runner,
+	capacity int,
+	runnerAvailableCh chan struct{}) runner.Runner {
+
+	qRunner := &QueueingRunner{
+		reqCh: make(chan interface{}),
+
+		delegate:  delegate,
+		capacity:  capacity,
+		nextRunId: 0,
+
+		runnerAvail: true,
+
+		q:       nil,
+		qToDel:  make(map[runner.RunId]runner.RunId),
+		delToQ:  make(map[runner.RunId]runner.RunId),
+		errored: make(map[runner.RunId]runner.ProcessStatus),
+	}
+
+	// push runner available notifications into our request channel
+	go func() {
+		for st := range runnerAvailableCh {
+			qRunner.reqCh <- st
+		}
+	}()
+
+	// start the request event processor
+	go qRunner.eventLoop()
+
+	return qRunner
+}
+
+// Implement Runner's methods by sending a request to qr.reqCh and waiting
+
+func (qr *QueueingRunner) Run(cmd *runner.Command) (runner.ProcessStatus, error) {
+	statusCh, errCh := make(chan runner.ProcessStatus), make(chan error)
```

Comment:
Discussed offline.

We don't need to per se, but I think it's easier to reason about so we don't have to convince ourselves.

I call this the "self-addressed, stamped envelope" pattern. A request includes the destination.